{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "Rtr15LT2UIZs"
   },
   "source": [
    " **IBM Virus Species Jump Hackathon at Deep Learning IndabaX 2019 in Durban**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "qiHH9xS8URXO"
   },
   "source": [
    "Goal: Predict whether an unknown animal virus could potentially infect humans given its genome (DNA or RNA) sequence.\n",
    "\n",
    "Importance:\n",
    "\n",
    "Many old and new dangerous viruses infecting humans emerge from animals. Emergence of new viruses is currently considered one of the biggest existential threats facing humanity. In 1918, a world-wide flu outbreak killed over 50 million people around the world. Ebola virus is thought to have come from bats, HIV virus is thought to have come from monkeys, SARS virus potentially came from birds, and many more. We can today sequence thousands of viruses in animals but it is hard to identify which of the tens of thousands of viruses present in wild and domestic animals could potentially cross over to infect humans. Thus, a computational model to predict whether an animal virus can infect humans would be of huge importance.\n",
    "\n",
    "Training Data: Genome sequences of 70 human viruses that can be easily transmitted from animals to humans, e.g. ebola (Class Label: Zoonotic) and sequences of another 70 viruses that cannot be easily transmitted from animals to humans (Class Label: Non-Zoonotic).\n",
    "\n",
    "The genome sequences are basically a string of 4 characters (AGCT) and the sequence of each virus ranges from 2000 letters to 10,000 characters. If you feel the training/test dataset is small, an option is to fragment each virus genome into smaller pieces which could easily create a training set of 5,000 to 10,000 samples in each class (i.e we can fragment each viral genome into 100 pieces).\n",
    "\n",
    "Test Data: 60 Genome sequences with the class label hidden from participants but provided to hackathon organizers (30 from each class).\n",
    "\n",
    "Evaluation: Participants will have to provide a deep learning model and their predictions for each provided test sample. Evaluation will be based on the follwing two criteria:\n",
    "\n",
    "    Most innovative model\n",
    "    Performance based on AUC/ precision-recall curves"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "RkEocJS8UaiN"
   },
   "source": [
    "Let's get started. Import all the necessary Python libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 34
    },
    "colab_type": "code",
    "id": "V8m3NsrAUQKv",
    "outputId": "cd1f0d15-d1b7-482b-9201-c1bc22ec89d3"
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import string\n",
    "import re\n",
    "import matplotlib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "Q975OWC7UlyW"
   },
   "source": [
    "Load the data uploaded to the current Colab instance under /content/sample_data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "NkzFw1CRUewn"
   },
   "outputs": [],
   "source": [
    "#with open(\"./content/NonZoonoticVirusesTrain.fasta\") as f:\n",
    "with open(\"./content/NonZoonoticVirusesStringent0_0_GlobalViromeProject.fasta\") as f:\n",
    "    NonZoo_raw_data  = f.read()\n",
    "    \n",
    "#with open(\"./content/ZoonoticVirusesTrain.fasta\") as g:\n",
    "with open(\"./content/ZoonoticVirusSequencesStringentGlobalViromeProject.fasta\") as g:\n",
    "    Zoo_raw_data  = g.read()\n",
    "    \n",
    "with open(\"./content/VirusesTestInput.fasta\") as h:\n",
    "    test_raw_data  = h.read()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "lOqC51yjUwGy"
   },
   "source": [
    "Preprocessing the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "CzxPC7IUUsUD"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "NotebookApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "NonZoodata = NonZoo_raw_data.split(\">\")\n",
    "Zoodata = Zoo_raw_data.split(\">\")\n",
    "testdata = test_raw_data.split(\">\")\n",
    "\n",
    "print(NonZoodata)\n",
    "print(Zoodata)\n",
    "print(testdata)\n",
    "\n",
    "# dump the epmpy string in position [0]\n",
    "NonZoodata2 = NonZoodata[1:]\n",
    "Zoodata2 = Zoodata[1:]\n",
    "testdata2 = testdata[1:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "WXpO1KSlVAkM"
   },
   "source": [
    "Separate names and sequences into two lists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "3RvzaTzaU4Cu"
   },
   "outputs": [],
   "source": [
    "NonZoo_new_data = [x.split(\"\\n\",1) for x in NonZoodata2]\n",
    "Zoo_new_data = [x.split(\"\\n\",1) for x in Zoodata2]\n",
    "test_new_data = [x.split(\"\\n\",1) for x in testdata2]\n",
    "\n",
    "NonZoo_id_name = []\n",
    "Zoo_id_name = []\n",
    "test_id_name = []\n",
    "NonZoo_genome_sequence = []\n",
    "Zoo_genome_sequence = []\n",
    "test_genome_sequence = []\n",
    "\n",
    "for x in NonZoo_new_data:\n",
    "    NonZoo_id_name.append(x[0])\n",
    "    NonZoo_genome_sequence.append(x[1])\n",
    "    \n",
    "for x in Zoo_new_data:\n",
    "    Zoo_id_name.append(x[0])\n",
    "    Zoo_genome_sequence.append(x[1])\n",
    "    \n",
    "for x in test_new_data:\n",
    "    test_id_name.append(x[0])\n",
    "    test_genome_sequence.append(x[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "tAo0ulVpVOlh"
   },
   "source": [
    "Separate into ID, Description and Sequence. From the data, you can just plit the ID using the first occurance of a \"white space\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "yy_b2MZ8VK4U"
   },
   "outputs": [],
   "source": [
    "def clean_id_names(text):\n",
    "    id_and_names = text.split(' ',1)\n",
    "    return id_and_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "MXqUJjTEVU-H"
   },
   "outputs": [],
   "source": [
    "ZooID_Description = []\n",
    "NonZooID_Description = []\n",
    "testID_Description = []\n",
    "for x in Zoo_id_name:\n",
    "    ZooID_Description.append(clean_id_names(x))\n",
    "for x in NonZoo_id_name:\n",
    "    NonZooID_Description.append(clean_id_names(x)) \n",
    "for x in test_id_name:\n",
    "    testID_Description.append(clean_id_names(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "YtmclZ08VYR-"
   },
   "outputs": [],
   "source": [
    "Zoo_ID = []\n",
    "Zoo_Description = []\n",
    "\n",
    "for x in ZooID_Description:\n",
    "    Zoo_ID.append(x[0])\n",
    "    Zoo_Description.append(x[1])\n",
    "    \n",
    "NonZoo_ID = []\n",
    "NonZoo_Description = []\n",
    "\n",
    "for x in NonZooID_Description:\n",
    "    NonZoo_ID.append(x[0])\n",
    "    NonZoo_Description.append(x[1])\n",
    "    \n",
    "test_ID = []\n",
    "test_Description = []\n",
    "\n",
    "for x in testID_Description:\n",
    "    test_ID.append(x[0])\n",
    "    test_Description.append(x[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "Gq1NK_AzVpIn"
   },
   "source": [
    "Clean the sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "3pmlU8AuVmiI"
   },
   "outputs": [],
   "source": [
    "def clean_text(text):\n",
    "    \n",
    "    remove_lower = lambda text: re.sub('[a-z]', '', text)\n",
    "    \n",
    "    text = remove_lower(text)\n",
    "    text = text.strip()\n",
    "    text = text.replace('\\n', '')\n",
    "    return text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "b5BcZVSuVtnD"
   },
   "outputs": [],
   "source": [
    "NonZoo_clean_sequences = []\n",
    "Zoo_clean_sequences = []\n",
    "test_clean_sequences = []\n",
    "for seq in NonZoo_genome_sequence:\n",
    "    NonZoo_clean_sequences.append(clean_text(seq))\n",
    "    \n",
    "for seq in Zoo_genome_sequence:\n",
    "    Zoo_clean_sequences.append(clean_text(seq))\n",
    "    \n",
    "for seq in test_genome_sequence:\n",
    "    test_clean_sequences.append(clean_text(seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "8X2yPX0cV5xZ"
   },
   "source": [
    "Now we create two lists for classification, zeros and ones."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "idqz4P1nVwn3"
   },
   "outputs": [],
   "source": [
    "def zerolistmaker(n):\n",
    "    listofzeros = [0] * n\n",
    "    return listofzeros\n",
    "\n",
    "def onelistmaker(n):\n",
    "    listofzeros = [1] * n\n",
    "    return listofzeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "wRyHTQtDV_ws"
   },
   "outputs": [],
   "source": [
    "Zoolabels = zerolistmaker(len(Zoo_clean_sequences)) #Zootonic viruses as Class 0.\n",
    "NonZoolabels = onelistmaker(len(NonZoo_clean_sequences)) #NonZootonic viruses as Class 1."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "qzlHdAiiWPsU"
   },
   "source": [
    "Now merge the data and the labels."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 204
    },
    "colab_type": "code",
    "id": "G87kBivSWEZP",
    "outputId": "3a3e8266-0d96-4893-ebab-51b633b8e8ad"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Description</th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_003466.1</td>\n",
       "      <td>Andes virus segment S, complete sequence</td>\n",
       "      <td>TAGTAGTAGACTCCTTGAGAAGCTACTGCTGCGAAAGCTGGAATGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_003468.2</td>\n",
       "      <td>Andes virus segment L, complete genome</td>\n",
       "      <td>TAGTAGTAGACTCCGGGATAGAAAAAGTTAGAAAAATGGAAAAGTA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_003467.2</td>\n",
       "      <td>Andes virus segment M, complete genome</td>\n",
       "      <td>TAGTAGTAGACTCCGCAAGAAGAAGCAAAAAATTAAAGAAGTGAGT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_009026.2</td>\n",
       "      <td>Bussuquara virus, complete genome</td>\n",
       "      <td>AGTATTTCTTCTGCGTGAGACCATTGCGACAGTTCGTACCGGTGAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_004211.1</td>\n",
       "      <td>Banna virus strain JKT-6423 segment 1, complet...</td>\n",
       "      <td>GTATTAAAAATTATCAACAAGGAATGGACATTCAAGAACAATTTGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID                                        Description  \\\n",
       "0  NC_003466.1           Andes virus segment S, complete sequence   \n",
       "1  NC_003468.2             Andes virus segment L, complete genome   \n",
       "2  NC_003467.2             Andes virus segment M, complete genome   \n",
       "3  NC_009026.2                  Bussuquara virus, complete genome   \n",
       "4  NC_004211.1  Banna virus strain JKT-6423 segment 1, complet...   \n",
       "\n",
       "                                           Sequences  Labels  \n",
       "0  TAGTAGTAGACTCCTTGAGAAGCTACTGCTGCGAAAGCTGGAATGA...       0  \n",
       "1  TAGTAGTAGACTCCGGGATAGAAAAAGTTAGAAAAATGGAAAAGTA...       0  \n",
       "2  TAGTAGTAGACTCCGCAAGAAGAAGCAAAAAATTAAAGAAGTGAGT...       0  \n",
       "3  AGTATTTCTTCTGCGTGAGACCATTGCGACAGTTCGTACCGGTGAG...       0  \n",
       "4  GTATTAAAAATTATCAACAAGGAATGGACATTCAAGAACAATTTGA...       0  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Zoo_data_frame = [list(x) for x in zip(Zoo_ID,Zoo_Description,Zoo_clean_sequences,Zoolabels)]\n",
    "NonZoo_data_frame = [list(x) for x in zip(NonZoo_ID,NonZoo_Description,NonZoo_clean_sequences,NonZoolabels)]\n",
    "test_data_frame = [list(x) for x in zip(test_ID,test_Description,test_clean_sequences)]\n",
    "trainlist = Zoo_data_frame + NonZoo_data_frame\n",
    "dataframe = pd.DataFrame(trainlist, columns = ['ID' , 'Description', 'Sequences', 'Labels'])\n",
    "dataframe.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 204
    },
    "colab_type": "code",
    "id": "AMUc8I_5WT68",
    "outputId": "e8343291-7357-4ada-cd0a-fc0b8e73b499"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Description</th>\n",
       "      <th>Sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_002728.1</td>\n",
       "      <td>Nipah virus, complete genome</td>\n",
       "      <td>ACCAAACAAGGGAGAATATGGATACGTTAAAATATATAACGTATTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_005283.1</td>\n",
       "      <td>Dolphin morbillivirus, complete genome</td>\n",
       "      <td>ACCAGACAAAGCTGGCTAGGGGTAGAATAACAGATAATGATAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_001498.1</td>\n",
       "      <td>Measles virus, complete genome</td>\n",
       "      <td>ACCAAACAAAGTTGGGTAAGGATAGATCAATCAATGATCATATTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_004148.2</td>\n",
       "      <td>Human metapneumovirus, complete genome</td>\n",
       "      <td>ACGCGAAAAAAACGCGTATAAATTAAGTTACAAAAAAACATGGGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_003266.2</td>\n",
       "      <td>Human adenovirus E, complete genome</td>\n",
       "      <td>CATCATCAATAATATACCTTATTTTTTTTGTGTGAGTTAATATGCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID                             Description  \\\n",
       "0  NC_002728.1            Nipah virus, complete genome   \n",
       "1  NC_005283.1  Dolphin morbillivirus, complete genome   \n",
       "2  NC_001498.1          Measles virus, complete genome   \n",
       "3  NC_004148.2  Human metapneumovirus, complete genome   \n",
       "4  NC_003266.2     Human adenovirus E, complete genome   \n",
       "\n",
       "                                           Sequences  \n",
       "0  ACCAAACAAGGGAGAATATGGATACGTTAAAATATATAACGTATTT...  \n",
       "1  ACCAGACAAAGCTGGCTAGGGGTAGAATAACAGATAATGATAAATT...  \n",
       "2  ACCAAACAAAGTTGGGTAAGGATAGATCAATCAATGATCATATTCT...  \n",
       "3  ACGCGAAAAAAACGCGTATAAATTAAGTTACAAAAAAACATGGGAC...  \n",
       "4  CATCATCAATAATATACCTTATTTTTTTTGTGTGAGTTAATATGCA...  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testframe = pd.DataFrame(test_data_frame, columns = ['ID' , 'Description', 'Sequences'])\n",
    "testframe.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 170
    },
    "colab_type": "code",
    "id": "ECjt3LYKWYMj",
    "outputId": "7e379e2d-866e-48a2-ea43-6fe1378ae88b"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID             0\n",
      "Description    0\n",
      "Sequences      0\n",
      "Labels         0\n",
      "dtype: int64\n",
      "ID             0\n",
      "Description    0\n",
      "Sequences      0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Check for any nulls values\n",
    "print(dataframe.isnull().sum())\n",
    "print(testframe.isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1969
    },
    "colab_type": "code",
    "id": "DDKGxX9zWd4n",
    "outputId": "0af01ac9-2bb7-4a50-bda6-0c6f9b44de22"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TAGTAGTAGACTCCTTGAGAAGCTACTGCTGCGAAAGCTGGAATGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TAGTAGTAGACTCCGGGATAGAAAAAGTTAGAAAAATGGAAAAGTA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TAGTAGTAGACTCCGCAAGAAGAAGCAAAAAATTAAAGAAGTGAGT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGTATTTCTTCTGCGTGAGACCATTGCGACAGTTCGTACCGGTGAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTATTAAAAATTATCAACAAGGAATGGACATTCAAGAACAATTTGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GTATAAATTTTGTATCCGATAGTAGCACATAGAATCACAATCACAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GTATTAAATTTTTGTAAGTCGTTATGGAATTATTTAGTGACAGTGG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GTATTTAAAATTCATGTTTTTGCATCATGGCGTGGGTTACGCAAGC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GTATTAAAAATTACAAGAACCTAACATTGCAATGGAGATCTTGAGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GTATTTAAAATTATAGAAAGTTCTGAACCTAGGGGTCTTTCTGTCT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GTATTAAAAATTGATAGCGCTTGCGTAAGAGTAAGATCTCAAGATG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GTATTTAAAATCATAGATTGAATATGTTATCAGAGACTGAGTTAAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GTATTTAAAATTGATATCAACCTTGACCAACAATGGCAAATAGAGC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GTATTAAAATTCAGCAATTGTCCAATTTAGGAAACATTCTGTTTAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GTATTTAAAATTGTGAGGAAAGACCAACATGGATGTACTGAGTAAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ATGGTTAATCCAAAGGGCGTCAATGTAATGGCAACCCGTGCAAAAC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GAACTCGGGAGGCGGTCGTGCGGACGCACGGACGCACGGACGCACG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AGTAGTGTACTCCACACTACAAACTTGCTATTGTTGAAAATCGCTG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AGTAGTGTACTACCGATACATCACAAACCTTTCAGAGACACATCTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AGTAGTGTACTCCTACATATAGAAAATTTAAAAATATAACCAGTAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ACCAGACAAAGTTGGCTAAGGATAGTTAAATTATTGAATATTTTAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CCCTCAGTTATACAATTACTAAAGTGCAAAAAAAGAACATAAAGGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GTTGGAGCAATGGCAATCGTTATTCTTCTAGCCCTGGTGGCATTAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CGGAGGTAAATAACCAGAGATGTCAGTGCCTACTTTTGAATTTACG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>GACCCGGGCCCTCCCCCGGCGGCGCGCGCGCGAAAATAAACGCGCG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ACGAAGACAAAAAAACCATTTAAACGATTATATAGTCGTACACGAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>GTAAATAAATCCTTTGATGTCTATAGCTGTTCTGCTCGTGCCTGGT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>GTAAAATTTCCATACGTATCTGACGCAGGATGGGTAAGATAATCAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>GTAAAATTATGCCTGAGCCTCACGCAGTAATCTATCTAACCAACGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>GTAAAATTCACTAGTCATGCTATCTGAGTTAGCTGCTCGTTTTGAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>ATGCCATACTGGTTGCCCACATATAACTTCGAAGGACTGCAATGCT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>ACCAAACAAGAGAAGAGACTTGCTTGGGAATATTAATTCAAATAAA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>ATATTTTTATTACTCATTTCCTGCATCCACCATATTGGCTGCCCGG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>TTTCCTCTTTTAGGGCGCGCCTCGGAGGCCTGTCCACTACTTCTGT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>ACGCGAAAAAATGCGTATAACAAACCTGTACATCCAAAAAAGATCG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>TGACGTATCTTTAGTTGATTTTGAAGCCTTGAAATAAAGCTTTTAA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>GTATACGAGAATTAGAAAAGGCACTCGTATACGTATTGGGCAATTA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>CGGATTGGATGTCGACAAATATTGAATTGGCAACGTAGGGAACCTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>AGTAGTGTACTACTAGATCCTAATCCTGTAAAATGGCAATGTTTGA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>AGTAGTGTACTCCACTTAAGTACTAAAGCTATTAATTGAGGACATC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>AGTAGTGTACTCCTATCTATACAGATACAAACAAACATAACCATGA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>ATGTCTAAAAAACCAGGAGGACCCGGCAAGAACCGGGTCGTCAATA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>TGACTAATTTACTTGTCTGGGCTGTGAACACCAGACTAGACGCAGG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>GGCTCAAGCCTCTCGCGGCTGTGCAAGTGGAAAAAAAATTTTTTTA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>CATCATCAATAATATACAGGACAAAGAGGTGTGGCCTAAATGTTGT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>GTTAATGAGAAATGGCTTCTGCCATCGCTCTCTCGAGCTCAACCGC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>CTCCATATAAAGACAATATCAAAATGGAGGACGGAAGTGTGATTCT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>AAAAGGTGTGTTCTCTTATTGTAGCTAACAACAATCTTACTTACAG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>TAGTAGACTCCTTGAGAAGCTACTACGACAAAGGCTGGAATGAGCA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>TAGTAGTAGACTCCGCACGAAGAAGCAAAAAAGTAGATAAAGGAAA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>GAGTTCTAGGAGAGTCCCTCCTAGTCTCTCCTCTCCGAGGAGGTAC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>CGCGGAGTCCCGCCGTACCGGTTTTGCCACCCCCGGGGGGCCTTTC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>CCCCTTAACAGGGGGGAGGGGGGGTTGAGGGGCCGGGATGGATGTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>AGTCCGGGCCCTCCCCCGGCGGCGCGCGCGCCGAAAAAAACGCGCG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>GATCGGGAAAACGCGAGGGGAGCGGGGGACAGGGGACGGCGTGTGC...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>GGGCCGGGCCCTCCCGCGGCGGCGCGCGCGCGGAAAAAAACGCCGG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>GGTCTCCGGGCCTTTCTCCCTGCGCTGGTCCCCCGGGGGGAGTTTG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>CGGGGGGGTGTTTGTGGCGGGGGGGTAAAATTTGGCCAGCTGCCAA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>GGCCGGGGGTCCCCCCGGGGGCAACGGGAGGAAAAAAAAACCGAGG...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>ACCAGACAAAGCTGGCTAGGGGTAGAATAACAGATAATGATAAATT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>201 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Sequences  Labels\n",
       "0    TAGTAGTAGACTCCTTGAGAAGCTACTGCTGCGAAAGCTGGAATGA...       0\n",
       "1    TAGTAGTAGACTCCGGGATAGAAAAAGTTAGAAAAATGGAAAAGTA...       0\n",
       "2    TAGTAGTAGACTCCGCAAGAAGAAGCAAAAAATTAAAGAAGTGAGT...       0\n",
       "3    AGTATTTCTTCTGCGTGAGACCATTGCGACAGTTCGTACCGGTGAG...       0\n",
       "4    GTATTAAAAATTATCAACAAGGAATGGACATTCAAGAACAATTTGA...       0\n",
       "5    GTATAAATTTTGTATCCGATAGTAGCACATAGAATCACAATCACAT...       0\n",
       "6    GTATTAAATTTTTGTAAGTCGTTATGGAATTATTTAGTGACAGTGG...       0\n",
       "7    GTATTTAAAATTCATGTTTTTGCATCATGGCGTGGGTTACGCAAGC...       0\n",
       "8    GTATTAAAAATTACAAGAACCTAACATTGCAATGGAGATCTTGAGA...       0\n",
       "9    GTATTTAAAATTATAGAAAGTTCTGAACCTAGGGGTCTTTCTGTCT...       0\n",
       "10   GTATTAAAAATTGATAGCGCTTGCGTAAGAGTAAGATCTCAAGATG...       0\n",
       "11   GTATTTAAAATCATAGATTGAATATGTTATCAGAGACTGAGTTAAG...       0\n",
       "12   GTATTTAAAATTGATATCAACCTTGACCAACAATGGCAAATAGAGC...       0\n",
       "13   GTATTAAAATTCAGCAATTGTCCAATTTAGGAAACATTCTGTTTAA...       0\n",
       "14   GTATTTAAAATTGTGAGGAAAGACCAACATGGATGTACTGAGTAAA...       0\n",
       "15   ATGGTTAATCCAAAGGGCGTCAATGTAATGGCAACCCGTGCAAAAC...       0\n",
       "16   GAACTCGGGAGGCGGTCGTGCGGACGCACGGACGCACGGACGCACG...       0\n",
       "17   AGTAGTGTACTCCACACTACAAACTTGCTATTGTTGAAAATCGCTG...       0\n",
       "18   AGTAGTGTACTACCGATACATCACAAACCTTTCAGAGACACATCTT...       0\n",
       "19   AGTAGTGTACTCCTACATATAGAAAATTTAAAAATATAACCAGTAG...       0\n",
       "20   ACCAGACAAAGTTGGCTAAGGATAGTTAAATTATTGAATATTTTAT...       0\n",
       "21   CCCTCAGTTATACAATTACTAAAGTGCAAAAAAAGAACATAAAGGA...       0\n",
       "22   GTTGGAGCAATGGCAATCGTTATTCTTCTAGCCCTGGTGGCATTAA...       0\n",
       "23   CGGAGGTAAATAACCAGAGATGTCAGTGCCTACTTTTGAATTTACG...       0\n",
       "24   GACCCGGGCCCTCCCCCGGCGGCGCGCGCGCGAAAATAAACGCGCG...       0\n",
       "25   ACGAAGACAAAAAAACCATTTAAACGATTATATAGTCGTACACGAA...       0\n",
       "26   GTAAATAAATCCTTTGATGTCTATAGCTGTTCTGCTCGTGCCTGGT...       0\n",
       "27   GTAAAATTTCCATACGTATCTGACGCAGGATGGGTAAGATAATCAA...       0\n",
       "28   GTAAAATTATGCCTGAGCCTCACGCAGTAATCTATCTAACCAACGA...       0\n",
       "29   GTAAAATTCACTAGTCATGCTATCTGAGTTAGCTGCTCGTTTTGAG...       0\n",
       "..                                                 ...     ...\n",
       "171  ATGCCATACTGGTTGCCCACATATAACTTCGAAGGACTGCAATGCT...       1\n",
       "172  ACCAAACAAGAGAAGAGACTTGCTTGGGAATATTAATTCAAATAAA...       1\n",
       "173  ATATTTTTATTACTCATTTCCTGCATCCACCATATTGGCTGCCCGG...       1\n",
       "174  TTTCCTCTTTTAGGGCGCGCCTCGGAGGCCTGTCCACTACTTCTGT...       1\n",
       "175  ACGCGAAAAAATGCGTATAACAAACCTGTACATCCAAAAAAGATCG...       1\n",
       "176  TGACGTATCTTTAGTTGATTTTGAAGCCTTGAAATAAAGCTTTTAA...       1\n",
       "177  GTATACGAGAATTAGAAAAGGCACTCGTATACGTATTGGGCAATTA...       1\n",
       "178  CGGATTGGATGTCGACAAATATTGAATTGGCAACGTAGGGAACCTT...       1\n",
       "179  AGTAGTGTACTACTAGATCCTAATCCTGTAAAATGGCAATGTTTGA...       1\n",
       "180  AGTAGTGTACTCCACTTAAGTACTAAAGCTATTAATTGAGGACATC...       1\n",
       "181  AGTAGTGTACTCCTATCTATACAGATACAAACAAACATAACCATGA...       1\n",
       "182  ATGTCTAAAAAACCAGGAGGACCCGGCAAGAACCGGGTCGTCAATA...       1\n",
       "183  TGACTAATTTACTTGTCTGGGCTGTGAACACCAGACTAGACGCAGG...       1\n",
       "184  GGCTCAAGCCTCTCGCGGCTGTGCAAGTGGAAAAAAAATTTTTTTA...       1\n",
       "185  CATCATCAATAATATACAGGACAAAGAGGTGTGGCCTAAATGTTGT...       1\n",
       "186  GTTAATGAGAAATGGCTTCTGCCATCGCTCTCTCGAGCTCAACCGC...       1\n",
       "187  CTCCATATAAAGACAATATCAAAATGGAGGACGGAAGTGTGATTCT...       1\n",
       "188  AAAAGGTGTGTTCTCTTATTGTAGCTAACAACAATCTTACTTACAG...       1\n",
       "189  TAGTAGACTCCTTGAGAAGCTACTACGACAAAGGCTGGAATGAGCA...       1\n",
       "190  TAGTAGTAGACTCCGCACGAAGAAGCAAAAAAGTAGATAAAGGAAA...       1\n",
       "191  GAGTTCTAGGAGAGTCCCTCCTAGTCTCTCCTCTCCGAGGAGGTAC...       1\n",
       "192  CGCGGAGTCCCGCCGTACCGGTTTTGCCACCCCCGGGGGGCCTTTC...       1\n",
       "193  CCCCTTAACAGGGGGGAGGGGGGGTTGAGGGGCCGGGATGGATGTT...       1\n",
       "194  AGTCCGGGCCCTCCCCCGGCGGCGCGCGCGCCGAAAAAAACGCGCG...       1\n",
       "195  GATCGGGAAAACGCGAGGGGAGCGGGGGACAGGGGACGGCGTGTGC...       1\n",
       "196  GGGCCGGGCCCTCCCGCGGCGGCGCGCGCGCGGAAAAAAACGCCGG...       1\n",
       "197  GGTCTCCGGGCCTTTCTCCCTGCGCTGGTCCCCCGGGGGGAGTTTG...       1\n",
       "198  CGGGGGGGTGTTTGTGGCGGGGGGGTAAAATTTGGCCAGCTGCCAA...       1\n",
       "199  GGCCGGGGGTCCCCCCGGGGGCAACGGGAGGAAAAAAAAACCGAGG...       1\n",
       "200  ACCAGACAAAGCTGGCTAGGGGTAGAATAACAGATAATGATAAATT...       1\n",
       "\n",
       "[201 rows x 2 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove unwanted features \n",
    "clean_data = dataframe.drop(['ID', 'Description'],1)\n",
    "clean_data.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1909
    },
    "colab_type": "code",
    "id": "g6OdXQ6LWx9V",
    "outputId": "6f6bdd82-48d8-4943-9207-ea4641c60474"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACCAAACAAGGGAGAATATGGATACGTTAAAATATATAACGTATTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACCAGACAAAGCTGGCTAGGGGTAGAATAACAGATAATGATAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACCAAACAAAGTTGGGTAAGGATAGATCAATCAATGATCATATTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACGCGAAAAAAACGCGTATAAATTAAGTTACAAAAAAACATGGGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CATCATCAATAATATACCTTATTTTTTTTGTGTGAGTTAATATGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TTAAAACAGCCTGTGGGTTGTTCCCACCCACAGGCCCATTGGGCGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CGCACCGGGGATCCTAGGCAATTTGGTTGTCTTTTTTTGAGGCCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GGCTCAAGCCTCTCGCGGCTGTGCAAGTGGAAAAAAAATTTTTTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AAAAGGTGTGTTCTCTTATTGTAGCTAACAACAATCTTACTTACAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CGCACCGGGGATCCTAGGCTTTTTGGATTGCGCTTTCCTCTAGATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CGGGGGGGTGTTTGTGGCGGGGGGGTAAAATTTGGCCAGCTGCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GTTAGTAAATTATATACATAATTTTATAATTAATTTAATTTTACTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GGTGGGGTGGGTACACTTCTCGCTCGATGAGTGGGGGTGTGGCTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ACCAAACAAGAGAAGAGACTTGTTTGGAAATATAAATTTAAATTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CATCATCAATAATATACCTTAAAACTGGAAACGAGCCAATATGATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GTAAAATTTTCGAATTTTATAATAATTAAGAAAGGAGTTTAATAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ACCAAGGGGAAAATGAAGTGGTGACTCAAATCATCGAAGACCCTCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CGGATTGGATGTCGACAAATATTGAATTGGCAACGTAGGGAACCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GCGCACAGTGGATCCTAGGCGATTTTGGTTACGCTATAATTGTAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TGACGTATCTTTAGTTGATTTTGAAGCCTTGAAATAAAGCTTTTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CCCCTTAACAGGGGGGAGGGGGGGTTGAGGGGCCGGGATGGATGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AGTCCGGGCCCTCCCCCGGCGGCGCGCGCGCCGAAAAAAACGCGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AGTAGTGTACTCCTATCTATACAGATACAAACAAACATAACCATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GTTAATGAGAAATGGCTTCTGCCATCGCTCTCTCGAGCTCAACCGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ACGCGAAAAAATGCGTATAACAAACCTGTACATCCAAAAAAGATCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CGCACCGAGGATCCTAGGCTTTTTGATGCGCAATGGATGAAATCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>GTAAATTAAATGCTATTTACAAAATTTAAACAGAAAGGAGAGATGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>GAGTTCTAGGAGAGTCCCTCCTAGTCTCTCCTCTCCGAGGAGGTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>GGCCGGGGGTCCCCCCGGGGGCAACGGGAGGAAAAAAAAACCGAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>TTTCCTCTTTTAGGGCGCGCCTCGGAGGCCTGTCCACTACTTCTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>AGTAGTGTACTCCACTTAAGTACTAAAGCTATTAATTGAGGACATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ATATTTTTATTACTCATTTCCTGCATCCACCATATTGGCTGCCCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>TGACTAATTTACTTGTCTGGGCTGTGAACACCAGACTAGACGCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>AGATTTTCTTGCACGTGCGATAGCTTCGGACAGCTTTGGCAGCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>CATCATCAATAATATACAGGACAAAGAGGTGTGGCCTAAATGTTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>GCGCACAGTGGATCCTAGGCGTAACTTCATCATTAAATCTCAGATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>CGCACAGTGGATCCTAGGCTATTGGATTGCGCTTTGCTTTTGTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>TAGTAGTAGACTCCGCACGAAGAAGCAAAAAAGTAGATAAAGGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>CATCATCAATAATATACCTTATTTTGGATTGAAGCCAATATGATAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>AGTAGTGTACTACTAGATCCTAATCCTGTAAAATGGCAATGTTTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>AGATTTTCTTGCACGTGCGTGCGCTTGCTTCAGACAGCAATAGCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>CTCCATATAAAGACAATATCAAAATGGAGGACGGAAGTGTGATTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>CATCATCAATAATATACCCCACAAAGTAAACAAAAGTTAATATGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>GGTCTCCGGGCCTTTCTCCCTGCGCTGGTCCCCCGGGGGGAGTTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>TTAAAACAGCTCTGGGGTTGTTCCCACCCCAGAGGCCCACGTGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>ACGCGAAAAAATGCGTACTACAAACTTGCACATTCGGAAAAAATGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>ACGCTTAACAACCAGATCAAAGAAGACACAGATAGTATCAGTGACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>ATGTCTAAAAAACCAGGAGGACCCGGCAAGAACCGGGTCGTCAATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>ATGCCATACTGGTTGCCCACATATAACTTCGAAGGACTGCAATGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>TCGCGGACCAGGAGTTCCTGCGGGAGGAGTTACAGCGGAGGCTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>GGGCCGGGCCCTCCCGCGGCGGCGCGCGCGCGGAAAAAAACGCCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>ACCAAACAAGAGAAGAGACTTGCTTGGGAATATTAATTCAAATAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>AGCCCGGGCCCCCCGCGGGCGCGCGCGCGCGCAAAAAAGGCGGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>AGACACACAAAAACAAGAGATGATGATTTTGTGTATCATATAAATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>TTAAAACAGCCTGTGGGTTGTACCCACCCACAGGGCCCACTGGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>TAGTAGACTCCTTGAGAAGCTACTACGACAAAGGCTGGAATGAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>CGCGGAGTCCCGCCGTACCGGTTTTGCCACCCCCGGGGGGCCTTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>GATCGGGAAAACGCGAGGGGAGCGGGGGACAGGGGACGGCGTGTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>GTATACGAGAATTAGAAAAGGCACTCGTATACGTATTGGGCAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>AGAATTTGTCTTGCTCTATTCACCGTTACTTTTCTTCTTGCCCGTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Sequences\n",
       "0   ACCAAACAAGGGAGAATATGGATACGTTAAAATATATAACGTATTT...\n",
       "1   ACCAGACAAAGCTGGCTAGGGGTAGAATAACAGATAATGATAAATT...\n",
       "2   ACCAAACAAAGTTGGGTAAGGATAGATCAATCAATGATCATATTCT...\n",
       "3   ACGCGAAAAAAACGCGTATAAATTAAGTTACAAAAAAACATGGGAC...\n",
       "4   CATCATCAATAATATACCTTATTTTTTTTGTGTGAGTTAATATGCA...\n",
       "5   TTAAAACAGCCTGTGGGTTGTTCCCACCCACAGGCCCATTGGGCGC...\n",
       "6   CGCACCGGGGATCCTAGGCAATTTGGTTGTCTTTTTTTGAGGCCTT...\n",
       "7   GGCTCAAGCCTCTCGCGGCTGTGCAAGTGGAAAAAAAATTTTTTTA...\n",
       "8   AAAAGGTGTGTTCTCTTATTGTAGCTAACAACAATCTTACTTACAG...\n",
       "9   CGCACCGGGGATCCTAGGCTTTTTGGATTGCGCTTTCCTCTAGATC...\n",
       "10  CGGGGGGGTGTTTGTGGCGGGGGGGTAAAATTTGGCCAGCTGCCAA...\n",
       "11  GTTAGTAAATTATATACATAATTTTATAATTAATTTAATTTTACTA...\n",
       "12  GGTGGGGTGGGTACACTTCTCGCTCGATGAGTGGGGGTGTGGCTCA...\n",
       "13  ACCAAACAAGAGAAGAGACTTGTTTGGAAATATAAATTTAAATTAA...\n",
       "14  CATCATCAATAATATACCTTAAAACTGGAAACGAGCCAATATGATA...\n",
       "15  GTAAAATTTTCGAATTTTATAATAATTAAGAAAGGAGTTTAATAGT...\n",
       "16  ACCAAGGGGAAAATGAAGTGGTGACTCAAATCATCGAAGACCCTCG...\n",
       "17  CGGATTGGATGTCGACAAATATTGAATTGGCAACGTAGGGAACCTT...\n",
       "18  GCGCACAGTGGATCCTAGGCGATTTTGGTTACGCTATAATTGTAAC...\n",
       "19  TGACGTATCTTTAGTTGATTTTGAAGCCTTGAAATAAAGCTTTTAA...\n",
       "20  CCCCTTAACAGGGGGGAGGGGGGGTTGAGGGGCCGGGATGGATGTT...\n",
       "21  AGTCCGGGCCCTCCCCCGGCGGCGCGCGCGCCGAAAAAAACGCGCG...\n",
       "22  AGTAGTGTACTCCTATCTATACAGATACAAACAAACATAACCATGA...\n",
       "23  GTTAATGAGAAATGGCTTCTGCCATCGCTCTCTCGAGCTCAACCGC...\n",
       "24  ACGCGAAAAAATGCGTATAACAAACCTGTACATCCAAAAAAGATCG...\n",
       "25  CGCACCGAGGATCCTAGGCTTTTTGATGCGCAATGGATGAAATCAT...\n",
       "26  GTAAATTAAATGCTATTTACAAAATTTAAACAGAAAGGAGAGATGT...\n",
       "27  GAGTTCTAGGAGAGTCCCTCCTAGTCTCTCCTCTCCGAGGAGGTAC...\n",
       "28  GGCCGGGGGTCCCCCCGGGGGCAACGGGAGGAAAAAAAAACCGAGG...\n",
       "29  TTTCCTCTTTTAGGGCGCGCCTCGGAGGCCTGTCCACTACTTCTGT...\n",
       "30  AGTAGTGTACTCCACTTAAGTACTAAAGCTATTAATTGAGGACATC...\n",
       "31  ATATTTTTATTACTCATTTCCTGCATCCACCATATTGGCTGCCCGG...\n",
       "32  TGACTAATTTACTTGTCTGGGCTGTGAACACCAGACTAGACGCAGG...\n",
       "33  AGATTTTCTTGCACGTGCGATAGCTTCGGACAGCTTTGGCAGCGGC...\n",
       "34  CATCATCAATAATATACAGGACAAAGAGGTGTGGCCTAAATGTTGT...\n",
       "35  GCGCACAGTGGATCCTAGGCGTAACTTCATCATTAAATCTCAGATT...\n",
       "36  CGCACAGTGGATCCTAGGCTATTGGATTGCGCTTTGCTTTTGTCAT...\n",
       "37  TAGTAGTAGACTCCGCACGAAGAAGCAAAAAAGTAGATAAAGGAAA...\n",
       "38  CATCATCAATAATATACCTTATTTTGGATTGAAGCCAATATGATAA...\n",
       "39  AGTAGTGTACTACTAGATCCTAATCCTGTAAAATGGCAATGTTTGA...\n",
       "40  AGATTTTCTTGCACGTGCGTGCGCTTGCTTCAGACAGCAATAGCAG...\n",
       "41  CTCCATATAAAGACAATATCAAAATGGAGGACGGAAGTGTGATTCT...\n",
       "42  CATCATCAATAATATACCCCACAAAGTAAACAAAAGTTAATATGCA...\n",
       "43  GGTCTCCGGGCCTTTCTCCCTGCGCTGGTCCCCCGGGGGGAGTTTG...\n",
       "44  TTAAAACAGCTCTGGGGTTGTTCCCACCCCAGAGGCCCACGTGGCG...\n",
       "45  ACGCGAAAAAATGCGTACTACAAACTTGCACATTCGGAAAAAATGG...\n",
       "46  ACGCTTAACAACCAGATCAAAGAAGACACAGATAGTATCAGTGACC...\n",
       "47  ATGTCTAAAAAACCAGGAGGACCCGGCAAGAACCGGGTCGTCAATA...\n",
       "48  ATGCCATACTGGTTGCCCACATATAACTTCGAAGGACTGCAATGCT...\n",
       "49  TCGCGGACCAGGAGTTCCTGCGGGAGGAGTTACAGCGGAGGCTGGA...\n",
       "50  GGGCCGGGCCCTCCCGCGGCGGCGCGCGCGCGGAAAAAAACGCCGG...\n",
       "51  ACCAAACAAGAGAAGAGACTTGCTTGGGAATATTAATTCAAATAAA...\n",
       "52  AGCCCGGGCCCCCCGCGGGCGCGCGCGCGCGCAAAAAAGGCGGGCG...\n",
       "53  AGACACACAAAAACAAGAGATGATGATTTTGTGTATCATATAAATA...\n",
       "54  TTAAAACAGCCTGTGGGTTGTACCCACCCACAGGGCCCACTGGGCG...\n",
       "55  TAGTAGACTCCTTGAGAAGCTACTACGACAAAGGCTGGAATGAGCA...\n",
       "56  CGCGGAGTCCCGCCGTACCGGTTTTGCCACCCCCGGGGGGCCTTTC...\n",
       "57  GATCGGGAAAACGCGAGGGGAGCGGGGGACAGGGGACGGCGTGTGC...\n",
       "58  GTATACGAGAATTAGAAAAGGCACTCGTATACGTATTGGGCAATTA...\n",
       "59  AGAATTTGTCTTGCTCTATTCACCGTTACTTTTCTTCTTGCCCGTT..."
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove unwanted features \n",
    "clean_test = testframe.drop(['ID', 'Description'],1)\n",
    "clean_test.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "mklddOzqW-SJ"
   },
   "source": [
    "# **This is where your model starts. Train your model on clean_data dataframe and predict classes for clean_test dataframe**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 204
    },
    "colab_type": "code",
    "id": "n8fwJg5NW3CR",
    "outputId": "cac5611b-1578-4367-ab93-dca62a912612"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TAGTAGTAGACTCCTTGAGAAGCTACTGCTGCGAAAGCTGGAATGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TAGTAGTAGACTCCGGGATAGAAAAAGTTAGAAAAATGGAAAAGTA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TAGTAGTAGACTCCGCAAGAAGAAGCAAAAAATTAAAGAAGTGAGT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGTATTTCTTCTGCGTGAGACCATTGCGACAGTTCGTACCGGTGAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTATTAAAAATTATCAACAAGGAATGGACATTCAAGAACAATTTGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Sequences  Labels\n",
       "0  TAGTAGTAGACTCCTTGAGAAGCTACTGCTGCGAAAGCTGGAATGA...       0\n",
       "1  TAGTAGTAGACTCCGGGATAGAAAAAGTTAGAAAAATGGAAAAGTA...       0\n",
       "2  TAGTAGTAGACTCCGCAAGAAGAAGCAAAAAATTAAAGAAGTGAGT...       0\n",
       "3  AGTATTTCTTCTGCGTGAGACCATTGCGACAGTTCGTACCGGTGAG...       0\n",
       "4  GTATTAAAAATTATCAACAAGGAATGGACATTCAAGAACAATTTGA...       0"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_data.head() # Training data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 204
    },
    "colab_type": "code",
    "id": "es08YYoaiMcS",
    "outputId": "17202f8a-bef4-41f5-c344-93454118ed39"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACCAAACAAGGGAGAATATGGATACGTTAAAATATATAACGTATTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACCAGACAAAGCTGGCTAGGGGTAGAATAACAGATAATGATAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACCAAACAAAGTTGGGTAAGGATAGATCAATCAATGATCATATTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACGCGAAAAAAACGCGTATAAATTAAGTTACAAAAAAACATGGGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CATCATCAATAATATACCTTATTTTTTTTGTGTGAGTTAATATGCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Sequences\n",
       "0  ACCAAACAAGGGAGAATATGGATACGTTAAAATATATAACGTATTT...\n",
       "1  ACCAGACAAAGCTGGCTAGGGGTAGAATAACAGATAATGATAAATT...\n",
       "2  ACCAAACAAAGTTGGGTAAGGATAGATCAATCAATGATCATATTCT...\n",
       "3  ACGCGAAAAAAACGCGTATAAATTAAGTTACAAAAAAACATGGGAC...\n",
       "4  CATCATCAATAATATACCTTATTTTTTTTGTGTGAGTTAATATGCA..."
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "pB_p0JbaXs1w"
   },
   "source": [
    "# **Make predictions on test data and save to file for submission**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "qq3-Bz19CQGy"
   },
   "outputs": [],
   "source": [
    "#clean_data = clean_data.drop(109)\n",
    "#clean_data = clean_data.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 357
    },
    "colab_type": "code",
    "id": "6aDRfYW8MlNl",
    "outputId": "e959c9fd-b861-485e-b639-9a1cdea5de26"
   },
   "outputs": [],
   "source": [
    "#clean_data['Sequences'][16]= clean_data['Sequences'][16][0:len(clean_data['Sequences'][16])-15]\n",
    "#clean_data['Sequences'][53]= (clean_data['Sequences'][53])[17:]\n",
    "#clean_data['Sequences'][54]= (clean_data['Sequences'][54])[16:]\n",
    "#clean_data['Sequences'][55]= (clean_data['Sequences'][55])[17:]\n",
    "\n",
    "def rotate(strg, n):\n",
    "    return strg[n:] + strg[:n]\n",
    "\n",
    "def feature_engineer(data):\n",
    "  protein = {\"TTT\" : \"F\", \"CTT\" : \"L\", \"ATT\" : \"I\", \"GTT\" : \"V\",\n",
    "             \"TTC\" : \"F\", \"CTC\" : \"L\", \"ATC\" : \"I\", \"GTC\" : \"V\",\n",
    "             \"TTA\" : \"L\", \"CTA\" : \"L\", \"ATA\" : \"I\", \"GTA\" : \"V\",\n",
    "             \"TTG\" : \"L\", \"CTG\" : \"L\", \"ATG\" : \"M\", \"GTG\" : \"V\",\n",
    "             \"TCT\" : \"S\", \"CCT\" : \"P\", \"ACT\" : \"T\", \"GCT\" : \"A\",\n",
    "             \"TCC\" : \"S\", \"CCC\" : \"P\", \"ACC\" : \"T\", \"GCC\" : \"A\",\n",
    "             \"TCA\" : \"S\", \"CCA\" : \"P\", \"ACA\" : \"T\", \"GCA\" : \"A\",\n",
    "             \"TCG\" : \"S\", \"CCG\" : \"P\", \"ACG\" : \"T\", \"GCG\" : \"A\",\n",
    "             \"TAT\" : \"Y\", \"CAT\" : \"H\", \"AAT\" : \"N\", \"GAT\" : \"D\",\n",
    "             \"TAC\" : \"Y\", \"CAC\" : \"H\", \"AAC\" : \"N\", \"GAC\" : \"D\",\n",
    "             \"TAA\" : \"Z\", \"CAA\" : \"Q\", \"AAA\" : \"K\", \"GAA\" : \"E\",\n",
    "             \"TAG\" : \"Z\", \"CAG\" : \"Q\", \"AAG\" : \"K\", \"GAG\" : \"E\",\n",
    "             \"TGT\" : \"C\", \"CGT\" : \"R\", \"AGT\" : \"S\", \"GGT\" : \"G\",\n",
    "             \"TGC\" : \"C\", \"CGC\" : \"R\", \"AGC\" : \"S\", \"GGC\" : \"G\",\n",
    "             \"TGA\" : \"Z\", \"CGA\" : \"R\", \"AGA\" : \"R\", \"GGA\" : \"G\",\n",
    "             \"TGG\" : \"W\", \"CGG\" : \"R\", \"AGG\" : \"R\", \"GGG\" : \"G\" \n",
    "             }\n",
    "\n",
    "\n",
    "\n",
    "  Features= []\n",
    "  counter=0 \n",
    "  for item in data['Sequences']:\n",
    "    #Basline DNA Sequencing\n",
    "    #print(counter)\n",
    "    counter+=1\n",
    "    DNA_length= len(item)\n",
    "    A=item.count(\"A\")/DNA_length\n",
    "    C=item.count(\"C\")/DNA_length\n",
    "    G=item.count(\"G\")/DNA_length\n",
    "    T=item.count(\"T\")/DNA_length\n",
    "\n",
    "\n",
    "    protein_length = DNA_length//3\n",
    "    #Protein Sequence/BOW\n",
    "    protein_sequence = \"\"\n",
    "    for i in range(0, len(item)-(3+len(item)%3), 3):\n",
    "      #if protein[item[i:i+3]] == \"Z\" :\n",
    "      #    break\n",
    "      \n",
    "      if item[i:i+3] in protein:\n",
    "        protein_sequence += protein[item[i:i+3]]\n",
    "      \n",
    "      \n",
    "\n",
    "    F=protein_sequence.count(\"F\")/protein_length\n",
    "    L=protein_sequence.count(\"L\")/protein_length\n",
    "    I=protein_sequence.count(\"I\")/protein_length\n",
    "    M=protein_sequence.count(\"M\")/protein_length\n",
    "    V=protein_sequence.count(\"V\")/protein_length\n",
    "    S=protein_sequence.count(\"S\")/protein_length\n",
    "    P=protein_sequence.count(\"P\")/protein_length\n",
    "    T=protein_sequence.count(\"T\")/protein_length\n",
    "    A=protein_sequence.count(\"A\")/protein_length\n",
    "    Y=protein_sequence.count(\"Y\")/protein_length\n",
    "    H=protein_sequence.count(\"H\")/protein_length\n",
    "    N=protein_sequence.count(\"N\")/protein_length\n",
    "    D=protein_sequence.count(\"D\")/protein_length\n",
    "    E=protein_sequence.count(\"E\")/protein_length\n",
    "    G=protein_sequence.count(\"G\")/protein_length\n",
    "    R=protein_sequence.count(\"R\")/protein_length\n",
    "    Q=protein_sequence.count(\"Q\")/protein_length\n",
    "    K=protein_sequence.count(\"K\")/protein_length\n",
    "    W=protein_sequence.count(\"W\")/protein_length\n",
    "    Z=protein_sequence.count(\"Z\")/protein_length\n",
    "    Features.append([A,C,G,T,DNA_length,F,L,I,M,V,S,P,T,A,Y,H,N,D,E,G,R,Q,K,W,Z])\n",
    "  return Features\n",
    "  \n",
    "  \n",
    "X = np.array(feature_engineer(clean_data) )   \n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.datasets import make_classification\n",
    "#X = np.array(Features)\n",
    "y = clean_data['Labels'].values\n",
    "#print(X.shape)\n",
    "clf = RandomForestClassifier(n_estimators=100, max_depth=5, random_state=0)\n",
    "clf.fit(X, y)\n",
    "pred = clf.predict(X)\n",
    "accuracy = (y==pred).sum() / len(y)\n",
    "#print(accuracy)\n",
    "\n",
    "\n",
    "###Validation Data Cleaning\n",
    "X_FINAL_TEST = np.array(feature_engineer(clean_test))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 68
    },
    "colab_type": "code",
    "id": "Qcjv_prnR6Ab",
    "outputId": "60f44669-e801-4d2c-e41d-c48bb6bb2de0"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: xgboost==0.80 in c:\\programdata\\anaconda3\\lib\\site-packages (0.80)\n",
      "Requirement already satisfied: numpy in c:\\programdata\\anaconda3\\lib\\site-packages (from xgboost==0.80) (1.15.1)\n",
      "Requirement already satisfied: scipy in c:\\programdata\\anaconda3\\lib\\site-packages (from xgboost==0.80) (1.0.0)\n"
     ]
    }
   ],
   "source": [
    "# latest version crashes on fit()\n",
    "!pip install xgboost==0.80  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 34
    },
    "colab_type": "code",
    "id": "wtXRt6lxTNqC",
    "outputId": "136a069f-dc55-45a2-ce5b-cbd7247b150a"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.5791904761904761 +/- 0.10128909480108697\n",
      "Min: 0.23809523809523808 Max: 0.9047619047619048\n"
     ]
    }
   ],
   "source": [
    "from xgboost import XGBClassifier\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.utils import shuffle\n",
    "import numpy as np\n",
    "\n",
    "# ignore deprecation warnings\n",
    "import warnings\n",
    "with warnings.catch_warnings():\n",
    "    warnings.filterwarnings(\"ignore\",category=DeprecationWarning)\n",
    "\n",
    "    test_acc = 0\n",
    "    rep = 1000\n",
    "    test_accs = []\n",
    "    for i in range(rep):\n",
    "      #X, y = shuffle(X, y)\n",
    "      X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)\n",
    "\n",
    "      clf =XGBClassifier(max_depth=5,min_child_weight=3, gamma=0.2, subsample=0.6, colsample_bytree=1.0)#, n_estimators=50)#, learning_rate=0.1, n_estimators=100)\n",
    "      clf.fit(X_train, y_train)\n",
    "      pred = clf.predict(X_train)\n",
    "      accuracy = (y_train==pred).sum() / len(y_train)\n",
    "    #   print(\"Train:\",accuracy)\n",
    "\n",
    "      pred = clf.predict(X_test)\n",
    "      accuracy = (y_test==pred).sum() / len(y_test)\n",
    "    #   print(\"Test:\",accuracy)\n",
    "      test_accs.append(accuracy)\n",
    "    test_accs = np.array(test_accs)\n",
    "    print(test_accs.mean(),\"+/-\",test_accs.std())\n",
    "    print(\"Min:\",test_accs.min(),\"Max:\",test_accs.max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 85
    },
    "colab_type": "code",
    "id": "vgJEACdBZgLr",
    "outputId": "336f3a3d-3ba7-4a30-b9bc-cb72f39fe9d1",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train: 0.945273631840796\n",
      "[0 1 0 0 0 0 0 1 1 0 1 0 0 0 1 0 0 1 0 1 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 0 0\n",
      " 0 0 1 0 1 0 1 0 0 0 1 1 0 1 1 0 0 0 0 1 1 1 0]\n",
      "60\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\preprocessing\\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\preprocessing\\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n"
     ]
    }
   ],
   "source": [
    "clf = XGBClassifier(max_depth=5,min_child_weight=3, gamma=0.2, subsample=0.6, colsample_bytree=1.0)\n",
    "clf.fit(X, y)\n",
    "pred = clf.predict(X)\n",
    "train_accuracy = (y==pred).sum() / len(y)\n",
    "print(\"Train:\",train_accuracy)\n",
    "y_test_pred = clf.predict(X_FINAL_TEST)\n",
    "print(y_test_pred)\n",
    "print(len(y_test_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "azmjQ5JwbSdK"
   },
   "outputs": [],
   "source": [
    "# y_pred = [str(x) for x in y_test_pred]\n",
    "# pred_y = \",\"\n",
    "# pred_y = pred_y.join(y_pred)\n",
    "# with open(\"./content/VirusesTestPrediction.fasta\",\"w+\") as o:\n",
    "#     o.write(pred_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "YMcjJ91XkPxh"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "IBMHackathon_XGBoost",
   "provenance": [],
   "toc_visible": true,
   "version": "0.3.2"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
